File: python/array_01/my_array.py
# -*- coding: utf-8 -*-

class MyArray:
    '''
    自定义一个int类型型数组，支持顺序插入，往指定位置插入、删除、按下标随机访问
    '''

    def __init__(self, capacity: int):
        self._data = []
        self._capacity = capacity

    def __getitem__(self, index: int) -> object:
        return self._data[index]

    def __setitem__(self, index: int, value: object):
        self._data[index] = value

    def __len__(self) -> int:
        return len(self._data)

    def add(self, value: int) -> bool:
        '''
        在数组尾部插入数据.
        :param value:  待插入的值
        :return:
        '''
        if len(self) >= self._capacity:
            return False
        else:
            self._data.append(value)
            return True

    def delete(self, index: int) -> bool:
        '''
         删除数组中指定下标的数据
        :param index: 下标
        :return:
        '''
        try:
            self._data.pop(index)
            return True
        except IndexError:
            return False

    def insert(self, index: int, value: int) -> bool:
        '''
         数组插入数据操作.
        :param index:  将要插入的下标
        :param value: 将要插入的数据
        :return: 插入成功返回True,否则返回False
        '''
        if len(self) >= self._capacity or index < 0:
            return False
        else:
            self._data.insert(index, value)
            return True

    def find(self, index: int) -> object:
        '''
        查找指定下标的数据
        :param index:  下标
        :return: 成功则返回找到的数据，失败返回False
        '''
        try:
            return self._data[index]
        except IndexError:
            return False

    def print_all(self):
        '''打印当前数组所有数据'''
        for value in self:
            print(value)


def test_myarray():
    array = MyArray(5)
    array.add(1)
    array.add(2)
    array.insert(2, 3)
    array.insert(2, 4)
    array.insert(2, 5)
    assert array.insert(0, 6) is False
    assert len(array) == 5
    assert array.find(2) == 5
    assert array.delete(4) is True
    array.print_all()


if __name__ == '__main__':
    test_myarray()

File: go/array_01/my_array.go
package array_01

import (
	"errors"
	"fmt"
)

type Array struct {
	data   [] int
	length int
}

// 初始化数组
func NewArray(capacity int) *Array {
	if capacity == 0 {
		return nil
	}
	return &Array{
		data:   make([]int, capacity, capacity),
		length: 0,
	}
}

//判断索引是否越界
func (a *Array) isIndexOutOfRange(index int) bool {
	if index >= a.length {
		return true
	}
	return false
}

//插入数值到索引index上
func (a *Array) Insert(index int, value int) error {
	if a.length == cap(a.data) {
		return errors.New("full array")
	}
	if index != a.length && a.isIndexOutOfRange(index) {
		return errors.New("out of index range")
	}
	// 如果当前插入位置不在最后，则需要把此位置后面的元素整体后移
	for i := a.length; i > index; i-- {
		a.data[i] = a.data[i-1]
	}
	a.data[index] = value
	a.length++
	return nil

}

// 从尾部插入
func (a *Array) InsertToTail(value int) error {
	return a.Insert(a.length, value)
}

//通过索引查找数组，索引范围[0,n-1]
func (a *Array) Find(index int) (int, error) {
	if a.isIndexOutOfRange(index) {
		return -1, errors.New("out of index range")

	}
	return a.data[index], nil
}

// 删除指定索引位置的值
func (a *Array) Delete(index int) error {
	if a.isIndexOutOfRange(index) {
		return errors.New("out of index range")
	}
	// 删除，后续元素前移
	for i := index; i < a.length-1; i++ {
		a.data[i] = a.data[i+1]
	}
	a.length--
	return nil
}

func (a *Array) PrintAll() {
	fmt.Println(a.data)
	//for i:=0;i<a.length;i++ {
	//
	//}
}
